src/Microsoft.ML.Vision/ImageClassificationTrainer.cs

```diff
@@ -1496,6 +1497,28 @@ public Classifier(ImageClassificationModelParameters model)
                 _runner.AddOutputs(model._graphOutputTensor);
                 _imageProcessor = new ImageClassificationTrainer.ImageProcessor(model._session,
                     model._imagePreprocessorTensorInput, model._imagePreprocessorTensorOutput);
+
+                try
+                {
+                    //Create a fake input and sess.run to mask the tensorflow graph initialization/
+                    // optimization time within Classifier creation, so that the first prediction using
+                    // Classifier.Score takes the same amount of time as the subsequent predictions.
+                    float[] fakeBatch = new float[224 * 224 * 3];
+                    long[] fakeTensorShape = { 1, 224, 224, 3 };
+                    var fakeBatchHandle = GCHandle.Alloc(fakeBatch, GCHandleType.Pinned);
+                    IntPtr fakeBatchPtr = fakeBatchHandle.AddrOfPinnedObject();
+                    int fakeBatchSizeInBytes = sizeof(float) * fakeBatch.Length;
+                    var fakeOutput = _runner.AddInput(new Tensor(fakeBatchPtr, fakeTensorShape, TF_DataType.TF_FLOAT, fakeBatchSizeInBytes), 0).Run();
```

Comment:
Are we sure doing this is a good idea overall?

I can think of the following places where this is actually worse for a user:

1. The total time take is now longer because we are making this "fake" prediction.
2. The user creates multiple PredictionEngines in their application - we will be issuing "fake" requests even for the 2nd, 3rd, etc.
3. The case where they create a PredictionEngine, but never make a prediction. (not very common, but still calling it out)

If some customers want this, they can "warm up" TensorFlow themselves. We shouldn't make every user pay.

I think the intention of the issue was to profile and investigate if there is something unnecessary happening, or if we could do something another way to speed it up. I'm not sure "masking" the time taken is a good solution.

Reply:
I agree with you @eerhardt. I haze been looking if TF has ways to warm things up but the one thing to try is to save graph in SaveModel format and see if that fixes this issue. Thanks for the investigation @ashbhandare but like @eerhardt says the point was to root cause the delay which you did but let’s try to not put hacks in.

Reply:
I agree that if there is a better way to do this, that should go in instead, but the issue specifically asked for masking this time as an alternative, which is why this PR exists. 
The fastest way of inferring in tensorflow from what I found is to freeze the model, apply certain graph optimizations like constant folding, remove unused nodes from the graph. This can be done by using the graph transform tools ([https://github.com/tensorflow/tensorflow/tree/master/tensorflow/tools/graph_transforms](https://github.com/tensorflow/tensorflow/tree/master/tensorflow/tools/graph_transforms)). I did not find a straightforward way of using these in ML .Net when we save our trained model.  

---
In reply to: [344231882](https://github.com/dotnet/machinelearning/pull/4456#discussion_r344231882) [](ancestors = 344231882)

Reply:
I would try with SaveModel format first before spending too much cycles on anything too fancy.

Reply:
SavedModel format is not supported by TF .Net. The one way that is supported is tf.train.saver.save(), which saves the checkpoint. Since we are freezing our trained model already(in this process, converting variable nodes to constants, and extracting the inference graph), I doubt that the saved model format would give much benefit(since the primary intution of using the saved_model format was that one can directly load the inference meta-graph https://www.tensorflow.org/guide/saved_model .)

Reply:
Run the sessions in a long-running process is a better for end-user. 
But there is another solution provided by TFX: https://www.tensorflow.org/tfx/serving/saved_model_warmup @codemzs @eerhardt

Reply:
@oceania2018 Thanks! @ashbhandare this is what I was talking about and yes, SaveModel is supported in TF .NET code, we have tests that exercise that code path, please check retrain tests.

Reply:
@codemzs I tried and failed to find a way of saving using the tensorflow-like api : https://www.tensorflow.org/api_docs/python/tf/saved_model/save, which produces a SavedModel with the following directory structure: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/saved_model/README.md#components. I might be missing something.